onnxruntime/core/graph/function.cc

```diff
@@ -470,15 +698,41 @@ FunctionImpl::FunctionImpl(const onnxruntime::Graph& graph,
         new_attr_map[(*node_attr).name()] = *node_attr;
       }
     }
-    function_body_graph.AddNode(uniq_identifier + "_" + std::to_string(node_index), (*node).op_type(),
+    function_body_graph.AddNode(uniq_identifier, (*node).op_type(),
                                 (*node).doc_string(), inputs, outputs, &new_attr_map, (*node).domain());
   }
 
   function_body_graph.SetInputs(graph_inputs);
   function_body_graph.SetOutputs(graph_outputs);
-  auto status = function_body_graph.Resolve();
 
-  ORT_ENFORCE(status.IsOK(), "Resolve subgraph failed:", status.ErrorMessage());
+  // Nested model local functions need to be initialized before Graph::Resolve() can be called for the function body.
+  // Parse the graph and initialize functions for nodes which reference model local functions.
+  // Only parse the graph if the model contains model local functions.
+  if (model_local_functions.size() > 0) {
+    for (auto node = function_body_graph.Nodes().begin(); node != function_body_graph.Nodes().end(); ++node) {
+      // Init nested functions
+      std::string func_identifier = node->Domain() + ":" + node->OpType();
+      auto iter = model_local_functions.find(func_identifier);
+      if (iter == model_local_functions.end()) {
+        continue;
+      }
+
+      // This node has a model local function proto.
+      auto onnx_function_proto = *(iter->second);
+      InitNestedModelLocalFunction(function_body_graph, node->Index(), onnx_function_proto, model_local_functions, function_container, logger, true);
+    }
+  }
+
+  // Graph resolve should be called on the parent functions only. Skip resolve if this is a nested function.
```

Comment:
Why is this distinction required? If we delay resolution for nested functions, it is unclear why we cannot do the same for non-nested functions. Is this meant to address model-local functions differently from schema-defined functions (which is a different dimension)? Or, it is just to preserve current behavior for existing models (to avoid unexpected surprises)?

Reply:
The idea is to do graph resolve top down starting from parent function and stepping into every nested function as we encounter it... This is needed for type inference to succeed.
We are treating model local functions different than schema-defined functions... Eventually I would like to change the current implementation to treat both schema-defined and model local functions uniformly... The change would simply make the code clean and simple... perf wise I don't think there is any cons in the current approach for schema-defined functions.